### DISCUSSION TITLE
Capi body read.

### DISCUSSION THREAD

**Discussion by gigagrig** [uNetworking/uWebSockets#1520] [Q&A] [answered]
Good afternoon!
I am trying to figure out how to use capi of uWebSockets for creating http server.
Apologies if I missed it, there is no way to read request body data, as far as i see.
It would be wonderful, if you assert my findings or point me to right way to use capi of uWebSockets.

Thanks in advance!

**Comment by gigagrig** [comment]
Question update : am I right that only the HttpResponse class has callbacks for reading  body data?
And with server app there is no possibility to read body or full request.

**Comment by cirospaciari** [comment, marked as answer]
You can read the body, grab the abort handler with uws_res_on_aborted and grab the data handler with uws_res_on_data, and grab the chunks, https://bun.sh/ and https://github.com/cirospaciari/socketify.py do this with this C API.

**Reply by gigagrig** [reply]
@cirospaciari thanks, a lot!
It was unclear that I must set uws_res_on_aborted handler first. It works now.
It is not clear when this handler is being invoked though.

**Reply by cirospaciari** [reply]
@gigagrig sorry that i have not seen you message for so long, just mention me when in need of help with C API is a pleasure to help.
